exp/etcdrestore/controllers/etcdmachinesnapshot_controller.go

```diff
@@ -60,18 +67,224 @@ func (r *EtcdMachineSnapshotReconciler) SetupWithManager(_ context.Context, mgr
+		// Snapshot creation is complete, no further action needed
+		return ctrl.Result{}, nil
+	}
+
+	// Requeue the request if necessary
+	if etcdMachineSnapshot.Status.Phase != snapshotrestorev1.ETCDSnapshotPhaseDone {
+		return ctrl.Result{RequeueAfter: 30 * time.Second}, nil
+	}
+
 	return ctrl.Result{}, nil
 }
 
 func (r *EtcdMachineSnapshotReconciler) reconcileDelete(
-	_ context.Context, _ *snapshotrestorev1.EtcdMachineSnapshot,
+	ctx context.Context, etcdMachineSnapshot *snapshotrestorev1.EtcdMachineSnapshot,
 ) (ctrl.Result, error) {
+	log := log.FromContext(ctx)
+
+	// Log the start of the deletion process
+	log.Info("Starting deletion of EtcdMachineSnapshot", "name", etcdMachineSnapshot.Name)
+
+	// Perform any necessary cleanup of associated resources here
+	// Example: Delete associated snapshot resources
+	snapshotList := &snapshotrestorev1.EtcdMachineSnapshotList{}
+	if err := r.Client.List(ctx, snapshotList, client.InNamespace(etcdMachineSnapshot.Namespace)); err != nil {
+		log.Error(err, "Failed to list associated EtcdMachineSnapshots")
+		return ctrl.Result{}, err
+	}
+
+	for _, snapshot := range snapshotList.Items {
+		if snapshot.Spec.MachineName == etcdMachineSnapshot.Spec.MachineName {
+			if err := r.Client.Delete(ctx, &snapshot); err != nil {
+				log.Error(err, "Failed to delete associated EtcdMachineSnapshot", "snapshotName", snapshot.Name)
+				return ctrl.Result{}, err
+			}
+			log.Info("Deleted associated EtcdMachineSnapshot", "snapshotName", snapshot.Name)
+		}
+	}
+
+	// Remove the finalizer to allow deletion of the EtcdMachineSnapshot
+	controllerutil.RemoveFinalizer(etcdMachineSnapshot, snapshotrestorev1.ETCDMachineSnapshotFinalizer)
+	if err := r.Client.Update(ctx, etcdMachineSnapshot); err != nil {
+		log.Error(err, "Failed to remove finalizer from EtcdMachineSnapshot")
+		return ctrl.Result{}, err
+	}
+
+	// Log the completion of the deletion process
+	log.Info("Completed deletion of EtcdMachineSnapshot", "name", etcdMachineSnapshot.Name)
+
 	return ctrl.Result{}, nil
 }
+
+// retryWithExponentialBackoff retries an operation with exponential backoff.
+func retryWithExponentialBackoff(ctx context.Context, initialDelay time.Duration, maxRetries int, operation func(ctx context.Context) error) error {
+	delay := initialDelay
+	for i := 0; i < maxRetries; i++ {
+		err := operation(ctx)
+		if err == nil {
+			return nil
+		}
+		fmt.Printf("Retry %d/%d failed: %v. Retrying in %v...", i+1, maxRetries, err, delay)
+		select {
+		case <-time.After(delay):
+			delay *= 2
+		case <-ctx.Done():
+			return ctx.Err()
+		}
+	}
+	return errors.New("operation failed after maximum retries")
+}
+
+// checkSnapshotStatus checks the status of the snapshot creation process.
+func checkSnapshotStatus(ctx context.Context, r *EtcdMachineSnapshotReconciler, etcdMachineSnapshot *snapshotrestorev1.EtcdMachineSnapshot) error {
+	log := log.FromContext(ctx)
+	etcdSnapshotFileList := &unstructured.UnstructuredList{}
```

Comment:
we should now use a wrapper library instead of unstructured

Reply:
Updated, PTAL